component/pyroscope/ebpf/ebpf_linux.go

```diff
@@ -0,0 +1,229 @@
+	CollectUserProfile   bool                   `river:"collect_user_profile,attr,optional"`
+	CollectKernelProfile bool                   `river:"collect_kernel_profile,attr,optional"`
+}
+
+func (rc *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*rc = defaultArguments()
+	type config Arguments
+	return f((*config)(rc))
+}
+
+func defaultArguments() Arguments {
+	return Arguments{
+		CollectInterval:      10 * time.Second,
+		SampleRate:           100,
+		PidCacheSize:         32,
+		ContainerIDCacheSize: 1024,
+		BuildIDCacheSize:     64,
+		SameFileCacheSize:    8,
+		CacheRounds:          3,
+		TargetsOnly:          true,
+		CollectUserProfile:   true,
+		CollectKernelProfile: true,
+	}
+}
+
+type Component struct {
+	options      component.Options
+	args         Arguments
+	argsUpdate   chan Arguments
+	appendable   *pyroscope.Fanout
+	targetFinder *sd.TargetFinder
+	session      *ebpfspy.Session
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	err := c.session.Start()
+	if err != nil {
+		return fmt.Errorf("ebpf profiling session start: %w", err)
+	}
+	defer c.session.Stop()
+
+	var g run.Group
+	g.Add(func() error {
+		collectInterval := c.args.CollectInterval
+		t := time.NewTicker(collectInterval)
+		defer t.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case newArgs := <-c.argsUpdate:
+				c.args = newArgs
+				c.updateTargetFinder()
+				c.session.UpdateCacheOptions(cacheOptionsFromArgs(c.args))
+				err := c.session.UpdateSampleRate(c.args.SampleRate)
+				if err != nil {
+					return nil
+				}
+				c.appendable.UpdateChildren(newArgs.ForwardTo)
+				if c.args.CollectInterval != collectInterval {
+					t.Reset(c.args.CollectInterval)
+					collectInterval = c.args.CollectInterval
+				}
+			case <-t.C:
+				var err error
+				pprof.Do(
+					context.Background(),
+					pprof.Labels("ebpf", "reset"),
+					func(ctx context.Context) {
+						err = c.reset()
+					})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}, func(error) {
+
+	})
+	return g.Run()
+}
+
+func cacheOptionsFromArgs(args Arguments) ebpfspy.CacheOptions {
+	return ebpfspy.CacheOptions{
+		PidCacheOptions: symtab.GCacheOptions{
+			Size:       args.PidCacheSize,
+			KeepRounds: args.CacheRounds,
+		},
+		BuildIDCacheOptions: symtab.GCacheOptions{
+			Size:       args.BuildIDCacheSize,
+			KeepRounds: args.CacheRounds,
+		},
+		SameFileCacheOptions: symtab.GCacheOptions{
+			Size:       args.SameFileCacheSize,
+			KeepRounds: args.CacheRounds,
+		},
+	}
+}
+
+func (c *Component) updateTargetFinder() {
+	c.targetFinder.SetTargets(sd.TargetsOptions{
+		Targets:       c.args.Targets,
+		DefaultTarget: c.args.DefaultTarget,
+		TargetsOnly:   c.args.TargetsOnly,
+	})
+	c.targetFinder.ResizeContainerIDCache(c.args.ContainerIDCacheSize)
+}
+
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+	c.argsUpdate <- newArgs
+	return nil
+}
+
+func (c *Component) reset() error {
```

Comment:
reset sounds like a component reset but in fact it sends data, may be we should find a better name.